File: android/src/main/java/com/payutesting/view/WebViewActivity.java

package com.payutesting.view;

import androidx.appcompat.app.AppCompatActivity;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.webkit.JavascriptInterface;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.Toast;

import com.payutesting.R;

import org.json.JSONException;
import org.json.JSONObject;

public class WebViewActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_webview);

        WebView wv1 = (WebView) findViewById(R.id.webView);
//        wv1.setWebViewClient(new MyBrowser());

        wv1.getSettings().setLoadsImagesAutomatically(true);
        wv1.getSettings().setJavaScriptEnabled(true);
        wv1.setScrollBarStyle(View.SCROLLBARS_INSIDE_OVERLAY);
        wv1.loadUrl(getIntent().getStringExtra("trans_url"));
        wv1.addJavascriptInterface(new WebAppInterface(this), "HTMLOUT");
        wv1.setWebViewClient(new WebViewClient() {
            @Override
            public boolean shouldOverrideUrlLoading(WebView view, String url) {
                view.loadUrl(url);
                return true;
            }
            @Override
            public void onPageStarted(WebView view, String url, Bitmap favicon) {
                super.onPageStarted(view, url, favicon);

            }
            @Override
            public void onPageFinished(WebView view, String url) {
                super.onPageFinished(view, url);
                Log.e("Redirect URL",""+url);
                if(url==getIntent().getStringExtra("redirect_url"))
                    view.loadUrl("javascript:window.HTMLOUT.processData(document.getElementById('result').value);");
            }

            @Override
            public void onReceivedError(WebView view, int errorCode, String description, String failingUrl) {
                Toast.makeText(getApplicationContext(), "Oh no! "+description , Toast.LENGTH_SHORT).show();
                super.onReceivedError(view, errorCode, description,failingUrl);
            }
        });
    }

    public void tetsme(View view) {
        Intent intentWithResult = new Intent();
        intentWithResult.putExtra("result", "Welcome");
        setResult(Activity.RESULT_OK, intentWithResult);
    }

    public class WebAppInterface {
        Context mContext;

        /** Instantiate the interface and set the context */
        WebAppInterface(Context c) {
            mContext = c;
        }

        @JavascriptInterface   // must be added for API 17 or higher
        public void processData(String toast) throws JSONException {
            // Log.d("Payment Response-->>>", "...>>>>>"+toast);
            //JSONObject object = new JSONObject(toast);
            Intent intentWithResult = new Intent();
            intentWithResult.putExtra("result", toast);
            setResult(Activity.RESULT_OK, intentWithResult);
            finish();
        }
    }

    @Override
    public void onBackPressed() {
//        super.onBackPressed();
        AlertDialog.Builder builder1 = new AlertDialog.Builder(WebViewActivity.this);
        builder1.setMessage("Are you sure you want to cancel?");
        builder1.setCancelable(true);

        builder1.setPositiveButton(
                "Cancel",
                new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int id) {
                        dialog.cancel();
                        sendErrorMessage("Payment failed");
                    }
                });

        builder1.setNegativeButton(
                "No",
                new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int id) {
                        dialog.cancel();
                    }
                });

        AlertDialog alert11 = builder1.create();
        alert11.show();

    }

    public void sendErrorMessage(String msg) {
       Intent intentWithResult= new Intent();
       intentWithResult.putExtra("ErrorType", "1" );
       intentWithResult.putExtra("ErrorMsg","Cancelled");
       setResult(Activity.RESULT_CANCELED,intentWithResult);
       finish();
    }

}

File: android/src/main/java/com/payutesting/view/WebViewActivity.kt
package com.payutesting.view

import android.annotation.SuppressLint
import android.app.AlertDialog
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.graphics.Bitmap
import android.app.Activity
import android.os.Build
import android.os.Bundle
import android.view.View
import android.webkit.*
import android.widget.Toast
import android.util.Log

import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDialog
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import com.payutesting.R
import com.payutesting.utilities.NetworkUtils
import com.payutesting.utilities.AvenuesParams

import kotlinx.android.synthetic.main.activity_webview.*

class WebViewActivity : AppCompatActivity() {

    private val networkUtils = NetworkUtils()
    private var transactionData : String? = null
    private var transactionURL : String? = null
    private var returnURL : String? = null
    private var cancelURL : String? = null
    private var intentWithResult : Intent  = Intent()

    companion object {
        private val TAG: String = "WebViewActivity"
    }

    override fun onStart() {
        super.onStart()

        LocalBroadcastManager.getInstance(this).registerReceiver(
            mNotificationReceiverInternet,
            IntentFilter(getString(R.string.keySendInternetStatus))
        )
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_webview)

        val webViewIntent : Intent = getIntent()
        transactionURL = webViewIntent.getStringExtra("paymenturl")


        
        if (networkUtils.haveNetworkConnection(this@WebViewActivity)) {
            loadWeb(transactionURL!!)
        } else {
            imgv_network_error.setVisibility(View.GONE)
            webView.setVisibility(View.VISIBLE)
            overlayView.visibility = View.VISIBLE
            sendErrorMessage("ERROR_NETWORK","No Network Connection")
        }
    }

    /**
     */
    @SuppressLint("SetJavaScriptEnabled", "AddJavascriptInterface", "ClickableViewAccessibility")
    private fun loadWeb(url: String) {
        val webSettings = webView.getSettings()
        webSettings.setJavaScriptEnabled(true)
        webSettings.setBuiltInZoomControls(false)
        webView.setWebViewClient(myWebClient())
        webView.addJavascriptInterface(JavaScriptHandler(), "HTMLOUT")
        try {
            transactionData?.toByteArray()?.let { webView.postUrl(url, it) }
        } catch (e: Exception) {
            sendErrorMessage("ERROR_GENERAL",e.message as String)
        }
    }

    /**
     *
     */
    inner class myWebClient : WebViewClient() {
        override fun onPageStarted(view: WebView, url: String, favicon: Bitmap?) {
            if (networkUtils.haveNetworkConnection(this@WebViewActivity)) {
                imgv_network_error.setVisibility(View.GONE)
                webView.setVisibility(View.VISIBLE)
                overlayView.visibility = View.VISIBLE
                super.onPageStarted(view, url, favicon)
            } else {
                webView.setVisibility(View.GONE)
                imgv_network_error.setVisibility(View.VISIBLE)
                overlayView.visibility = View.VISIBLE
                sendErrorMessage("ERROR_NETWORK","No Network Connection")
            }
        }

        override fun onPageFinished(view: WebView, url: String) {
            if (networkUtils.haveNetworkConnection(this@WebViewActivity)) {
                webView.setVisibility(View.VISIBLE)
                overlayView.visibility = View.GONE
                if (url.indexOf(returnURL as String) != -1) {
                    webView.loadUrl("javascript:window.HTMLOUT.processData(document.getElementById('result').value);");
                }
                super.onPageFinished(view, url)
            }else
                sendErrorMessage("ERROR_NETWORK","No Network Connection")
        }

        override fun onReceivedError(view: WebView, request: WebResourceRequest, error: WebResourceError) {
            try {
                webView.setVisibility(View.GONE)
                imgv_network_error.setVisibility(View.VISIBLE)
                overlayView.visibility = View.VISIBLE
            } catch (e: Exception) {
                sendErrorMessage("ERROR_GENERAL",e.message as String)
            }

        }
    }


    inner class JavaScriptHandler internal constructor() {
        @JavascriptInterface
        fun processData(result: String) {
            Log.d(TAG, result)
            intentWithResult.putExtra("result", result)
            setResult(RESULT_OK, intentWithResult)
            finish()
        }
    }

    /**
     * Back press callback onBackPressed
     */
    override fun onBackPressed() {
        generalDailog("CCAvenue", "Are you sure you want to quit?")
    }

    /**
     * Back Press Alert Dialog
     */
    fun generalDailog(title: String, message: String) {
        try {
            val builder = AlertDialog.Builder(this@WebViewActivity)

            builder.setTitle(title)
            builder.setMessage(message)
            builder.setCancelable(false)
            builder.setPositiveButton("YES") { _, _ ->
                try {
                    sendErrorMessage("ERROR_USER_CANCELLED","User Cancelled Transaction")
                } catch (e: Exception) {
                    sendErrorMessage("ERROR_GENERAL",e.message as String)
                }
            }
            builder.setNegativeButton("No") { dialog, _ ->
                dialog.cancel()
            }
            val dialog: AlertDialog = builder.create()
            dialog.show()
        } catch (e: Exception) {
            sendErrorMessage("ERROR_GENERAL",e.message as String)
        }
    }

    fun sendErrorMessage(errorType : String,errorMsg : String){
        intentWithResult.putExtra("ErrorType",errorType )
        intentWithResult.putExtra("ErrorMsg",errorMsg)
        setResult(RESULT_CANCELED,intentWithResult)
        finish()
    }


    
     
    private val mNotificationReceiverInternet = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent?) {

            if (intent != null && intent.extras != null && !intent.extras!!.isEmpty) {
                if (!intent.getBooleanExtra("isConnected", false)) {
                    showToast("No Internet Connection")
                    sendErrorMessage("ERROR_NETWORK","No Network Connection")
                }
            }
        }
    }

    private fun unRegisterListener(){
         try {
            LocalBroadcastManager.getInstance(this).unregisterReceiver(mNotificationReceiverInternet)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    /**
     */
    private fun showToast(text: String) {
        Toast.makeText(this, text, Toast.LENGTH_LONG).show()
    }

    

    /**
     *
     */
    override fun onDestroy() {
        unRegisterListener()
        super.onDestroy()
    }

}